dockstore-webservice/src/main/java/io/dockstore/webservice/languages/CWLHandler.java

```diff
@@ -896,4 +908,172 @@ public VersionTypeValidation validateToolSet(Set<SourceFile> sourcefiles, String
+        private static final List<String> MIXIN_KEYS = Arrays.asList("$mixin", "mixin");
+        private static final int DEFAULT_MAX_DEPTH = 8;
+
+        private final Set<SourceFile> secondarySourceFiles;
+        private final int maxDepth;
+        private final Map<String, String> idToPath;
+
+        public Preprocessor(Set<SourceFile> secondarySourceFiles, int maxDepth) {
+            this.secondarySourceFiles = secondarySourceFiles;
+            this.maxDepth = maxDepth;
+            this.idToPath = new HashMap<>();
+        }
+
+        public Preprocessor(Set<SourceFile> secondarySourceFiles) {
+            this(secondarySourceFiles, DEFAULT_MAX_DEPTH);
+        }
+
+        private void preprocessMapValues(Map<String, Object> map, String currentPath, int depth) {
+            map.replaceAll((k, v) -> preprocess(v, currentPath, depth));
+            Object runValue = map.get("run");
+            if (runValue instanceof String) {
+                LOG.error("RUN " + currentPath + " " + depth + ": " + runValue);
+                String runPath = (String)runValue;
+                map.put("run", loadFileAndPreprocess(resolvePath(runPath, currentPath), depth));
+            }
+        }
+
+        private void preprocessListValues(List<Object> list, String currentPath, int depth) {
+            list.replaceAll(v -> preprocess(v, currentPath, depth));
+        }
+
+        private boolean isTool(Map<String, Object> map) {
+            return Objects.equals(map.get("class"), "CommandLineTool");
+        }
+
+        private String addId(Map<String, Object> map) {
+            map.putIfAbsent("id", java.util.UUID.randomUUID().toString());
+            return (String)map.get("id"); // TODO fix?
+        }
+
+        public Object preprocess(Object obj, String currentPath, int depth) {
+
+            LOG.error("PREPROCESS " + currentPath + " " + depth + ": " + obj);
+
+            if (depth > maxDepth) {
+                error("maximum file depth exceeded");
+                return obj;
+            }
+
+            if (obj instanceof Map) {
+
+                Map map = (Map<String, Object>)obj;
+
+                if (isTool(map)) {
+                    idToPath.put(addId(map), currentPath);
+                }
+
+                String importPath = findValue(IMPORT_KEYS, map);
+                if (importPath != null) {
+                    LOG.error("import " + importPath + " " + currentPath);
+                    return loadFileAndPreprocess(resolvePath(importPath, currentPath), depth);
+                }
+
+                String includePath = findValue(INCLUDE_KEYS, map);
+                if (includePath != null) {
+                    LOG.error("include " + includePath + " " + currentPath);
+                    return loadFile(resolvePath(includePath, currentPath));
+                }
+
+                String mixinPath = findValue(MIXIN_KEYS, map);
```

Comment:
As a warning, not sure this will work. 
I'm not sure we're actually ingesting mixins properly but you can leave this in for when we do start ingesting them properly. 
I guess the testing below will shed better light on what works and what doesn't when fleshed out